### ISSUE TITLE
multiple optimizers

### ISSUE THREAD

**Issue by glample** [NVIDIA/apex#264]
Hi,

I'm working with the following setting where I have:
- a model on fp16 with fp16 apex optimizer O1
- a float32 parameter which is used in the forward path of my model, with a non-apex optimizer

I'm wondering how I could do the optimization here. From what I understand, I cannot do a regular O2.step() because the parameters in O2 will be scaled by the loss scaling coefficient, right? Is there a way I can extract this coefficient from O1 and downscale manually my gradients in O2 before calling O2.step(), or something like that?

Thanks

**Comment by mcarilli** [comment]
What do you mean by a "non-apex optimizer O2"?  O2 is specific to Apex (amp).

**Comment by glample** [comment]
Sorry I just meant a non-apex optimizer here. It's a bit related to https://github.com/NVIDIA/apex/issues/243: overall I want to do sparse updates on a float32 parameter, and have the rest of the model on fp16, but since apex does not support sparse I want to perform the updates manually on my float32 tensor (using regular sparse PyTorch optimizer)

But I'm afraid about mixing apex/non-apex optimizers, and that something unwanted happens during training, like the gradients of my sparse parameters are scaled so I have to unscale them before doing the update, or if I should not to any update at all because some overflow was detected, and my non-apex optimizer will not know it, or something like that. What do you think?

**Comment by mcarilli** [comment]
In general, I don't want people to run optimizers that haven't been passed through amp.initialize.  We plan to support sparse gradients in Amp soon.  Closing because https://github.com/NVIDIA/apex/pull/252 and https://github.com/NVIDIA/apex/issues/243 are tracking this.
